### PR TITLE
Prevent from inserting too many "end"s in Elixir

### LINKED ISSUE #8030: Too many closing "end"s in Elixir
#### Description :octocat:
Since upgrading spacemacs recently, too many closing `end` keywords are being inserted automatically in the Elixir language.

#### Reproduction guide :beetle:
- Start Emacs
- Enter elixir-mode, and make sure smartparens-mode is enabled
- Start a case statement with multiple clauses (see behaviours below)
- As you enter newlines, notice that too many `end`s are created

*Observed behaviour:* :eyes: :broken_heart:
```elixir
case something do
  {:ok, x} ->
    x
  {:ok, y} ->
    y
  {:ok, z} ->
    z
end
end
end
end
```

*Expected behaviour:* :heart: :smile:
```elixir
case something do
  {:ok, x} ->
    x
  {:ok, y} ->
    y
  {:ok, z} ->
    z
end
```

#### System Info :computer:
- OS: darwin
- Emacs: 25.1.1
- Spacemacs: 0.200.5
- Spacemacs branch: master (rev. 664ba6a)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: ivy
- Layers:
```elisp
(ansible auto-completion better-defaults clojure colors deft docker elixir elm erlang emacs-lisp git github haskell html ivy javascript markdown org osx python react restclient rust shell spell-checking syntax-checking themes-megapack version-control yaml)
```
- System configuration features: JPEG RSVG IMAGEMAGICK NOTIFY ACL GNUTLS LIBXML2 ZLIB TOOLKIT_SCROLL_BARS NS

### PR DESCRIPTION
**Pull request by michalmuskala**
Closes #8030

### CODE REVIEW

File: layers/+lang/elixir/funcs.el [hunk 1]
```diff
@@ -9,11 +9,18 @@
...
+    (while (looking-back " ") (backward-char))
+    (looking-back expr)))
+
 (defun spacemacs//elixir-do-end-close-action (id action context)
   (when (eq action 'insert)
-    (newline-and-indent)
-    (forward-line -1)
-    (indent-according-to-mode)))
+    (if (spacemas//elixir-looking-back-spaces-p id)
+        (progn (insert " ") (backward-char))
```
Comment (sdwolf) [inline layers/+lang/elixir/funcs.el]: :heart:

**Comment by michalmuskala** [comment]
I'll think about the other two cases today evening. Maybe there's a simple solution lurking around 😃

**Comment by d12frosted** [comment]
@michalmuskala great, let me know once it's ready for merge 😸 or is it irrelevant to this PR?

**Comment by michalmuskala** [comment]
Let's say it's a `[WIP]` for now. @d12frosted, I'll ping you in the evening when I'll know more.

**Comment by d12frosted** [comment]
@michalmuskala thanks!

**Comment by michalmuskala** [comment]
I tried to make it work with both `-` and `(` after `fn`. The current behaviour is:

```
fn| # type -
fn-| end
```
```
fn| # type (
fn(|) end
```

The `-` case was a simple one - we do exactly the same thing that for the `SPC` case. The parens is more complicated. I achieved the current behaviour by preventing the regular parens autocomplete and inserting the matching `)` manually after `fn`, when `(` was typed. This allows easily controlling the position of the point and prevents the `)` being inserted after the `end`.

My elisp is very basic, so if the code itself can be improved I'm open for suggestions.

@d12frosted provided there are no other suggestions, I think this is ready.

**Comment by d12frosted** [comment]
@michalmuskala Thank you :+1: Cherry-picked into `develop`, you can safely delete your branch.

This looks good to me. And I've played a bit with your PR - seems to fix all described issues. The only thing I have noticed is a minor lag when you type `case something do SPC`. But I am on a slow computer right now and everything lags, so I guess this is fault of my computer.

**Comment by pdilyard** [comment]
@michalmuskala Is there a way to ignore this in comments? I just noticed that if I type "do" in a comment, an "end" will be added when I hit the spacebar.

```
# The below function will do something end
```

**Pull request closed without merging**